Detectors/TPC/workflow/include/TPCWorkflow/TPCSectorCompletionPolicy.h

```diff
@@ -0,0 +1,164 @@
+/// stack describing the current sector and defining the active sectors in the setup. The completion
+/// policy callback will wait until there is data for all active sectors.
+///
+/// Parameters:
+///   processor name   rexexp to match a name of the processor for which the policy should be applied
+///   input matchers   provided as an argument pack
+///                    Note: it is important to use ConcreteDataTypeMatcher to define input spec with
+///                          wildcard on subSpecification
+/// Usage:
+///   TPCSectorCompletionPolicy("processor-name-regexp", InputSpec{"", ConcreteDataTypeMatcher{"DET", "RAWDATA"}}, ...)();
+///
+class TPCSectorCompletionPolicy
+{
+ public:
+  TPCSectorCompletionPolicy() = delete;
+  template <typename... Args>
+  TPCSectorCompletionPolicy(const char* processorName, Args&&... args)
+    : mProcessorName(processorName), mInputMatchers()
+  {
+    init(std::forward<Args>(args)...);
+  }
+
+  o2::framework::CompletionPolicy operator()()
+  {
+    constexpr static size_t NSectors = o2::tpc::Sector::MAXSECTOR;
+
+    auto matcher = [expression = mProcessorName](DeviceSpec const& device) -> bool {
+      return std::regex_match(device.name.begin(), device.name.end(), std::regex(expression.c_str()));
+    };
+
+    auto callback = [inputMatchers = mInputMatchers](CompletionPolicy::InputSet inputs) -> CompletionPolicy::CompletionOp {
+      auto op = CompletionPolicy::CompletionOp::Wait;
+      std::bitset<NSectors> validSectors = 0;
+      bool haveMatchedInput = false;
+      uint64_t activeSectors = 0;
+      size_t nActiveInputRoutes = 0;
+      size_t nMaxPartsPerRoute = 0;
+      int inputType = -1;
+      for (auto it = inputs.begin(), end = inputs.end(); it != end; ++it) {
+        nMaxPartsPerRoute = it.size() > nMaxPartsPerRoute ? it.size() : nMaxPartsPerRoute;
+        bool haveActivePart = false;
+        for (auto const& ref : it) {
+          if (!DataRefUtils::isValid(ref)) {
+            continue;
+          }
+          haveActivePart = true;
+          auto const* dh = DataRefUtils::getHeader<o2::header::DataHeader*>(ref);
+          // check if the O2 message matches on of the input specs to be matched and
+          // if it matches, check for the sector header, retrieve the active sector information
+          // and mark the sector as valid
+          for (size_t idx = 0, end = inputMatchers.size(); idx < end; idx++) {
+            auto const& spec = inputMatchers[idx];
+            if (DataRefUtils::match(ref, spec)) {
+              haveMatchedInput = true;
+              if (inputType == -1) {
+                // we bind to the index of the first match and require all other inputs to match the same spec
+              } else if (inputType != idx) {
+                std::stringstream error;
+                error << "routing error, input messages must all be of the same type previously bound to "
```

Comment:
I am not sure if we want to require in all cases that there is only one input. I'd add a flag to the constructor, such that this can be enforced or not. But OK, for now I think it is OK, we can also leave it as is and add such a flag later (It would make the logic a bit more compley, since you'd need activeSectors and validSectors per inputType, but of course this can be done).

Reply:
I think for the moment we should enforce this. Supporting mixed inputs would mean to check for every type separately and apply and AND logic. Lets keep it simple for now and extend if necessary. I'm planning anyhow to change the creation of the completion policy depending on the runtime configuration. Now we need to define two input types because the policy is created before everything else.